### PR TITLE
POC add Delimiter support for ListObjectsV2

### PR DESCRIPTION
**Pull request by dmolina79**
TODO: need more testing with FE to validate this solves their use case

## Description
Add Basic Delimiter Support to s3x for listObjectsV2

## Motivation and Context
Missing parity with s3 API for listObjectsV2

## How to test this PR?
const s3params = {
      Bucket: 'my-bucket',
      MaxKeys: 20,
      Prefix: 'folder/',
      Delimiter: '/',
    };

    console.log('s3params', s3params);
    s3.listObjectsV2(s3params, (err, data) => {
      if (err) {
        reject (err);
      }
      resolve (data);
    });

### CODE REVIEW

File: cmd/gateway/s3x/gateway-s3x-object.go [hunk 1]
```diff
@@ -49,10 +49,18 @@ func (x *xObjects) ListObjectsV2(
 	if err != nil {
 		return loi, x.toMinioErr(err, bucket, "", "")
 	}
-	loi.Objects = make([]minio.ObjectInfo, 0, len(objs))
-	for _, obj := range objs {
-		loi.Objects = append(loi.Objects, getMinioObjectInfo(&obj))
+
+	if delimiter != "" {
```
Comment (jsonsivar) [inline cmd/gateway/s3x/gateway-s3x-object.go:53]: Should this check for null as well?

File: cmd/gateway/s3x/gateway-s3x-object.go [hunk 2]
```diff
@@ -319,11 +327,35 @@ func (x *xObjects) emulateCreateFolder(
...
 
+// helper function for applying delimiter
+func filterObjectsWithDelimiter(objs []ObjectInfo, prefix string, delimiter string) ([]minio.ObjectInfo, []string) {
+	set := make(map[string]struct{})
+	prefixLen := len(prefix)
+	objects := make([]minio.ObjectInfo, 0, len(objs))
+	prefixes := make([]string, 0)
+	for _, obj := range objs {
+		sub := obj.Name[prefixLen:]
+		if idx := strings.Index(sub, delimiter); idx >= 0 {
```
Comment (jsonsivar) [inline cmd/gateway/s3x/gateway-s3x-object.go:338]: What happens if there is a delimiter after the length of the prefix? Assuming we want to break those down into the `prefixes` array as well or maybe that's not allowed input?

File: cmd/gateway/s3x/gateway-s3x-object.go [hunk 3]
```diff
@@ -319,11 +327,35 @@ func (x *xObjects) emulateCreateFolder(
...
+			// if we found delimiter we add prefix and filter object
+			currPrefix := sub[:idx]+delimiter
+			if _, ok := set[currPrefix]; !ok {
+				// only add elements that we dont have repeated
```
Comment (jsonsivar) [inline cmd/gateway/s3x/gateway-s3x-object.go:342]: If my understanding the requirements is correct, would have expected another `for` or recursion to add the nested folder structure inside each path from the `delimiter` or does it just work for one level?

File: cmd/gateway/s3x/gateway-s3x-object.go (hunk 1 above)
Comment (dmolina79) [reply cmd/gateway/s3x/gateway-s3x-object.go:53]: the delimiter is empty string by default, nil only applies for pointers

File: cmd/gateway/s3x/gateway-s3x-object.go (hunk 3 above)
Comment (dmolina79) [reply cmd/gateway/s3x/gateway-s3x-object.go:342]: just one level, from the s3 docs

File: cmd/gateway/s3x/gateway-s3x-object.go (hunk 2 above)
Comment (dmolina79) [reply cmd/gateway/s3x/gateway-s3x-object.go:338]: not understanding what you mean, but yeah, we strip down the prefix out of the name, and look for the delimiter in the rest of the string to find what we will include in the prefixes array

File: cmd/gateway/s3x/gateway-s3x-object.go (hunk 1 above)
Comment (jsonsivar) [reply cmd/gateway/s3x/gateway-s3x-object.go:53]: Ah, thank you.

File: cmd/gateway/s3x/gateway-s3x-object.go (hunk 3 above)
Comment (jsonsivar) [reply cmd/gateway/s3x/gateway-s3x-object.go:342]: Ok makes sense.

File: cmd/gateway/s3x/gateway-s3x-object.go (hunk 2 above)
Comment (jsonsivar) [reply cmd/gateway/s3x/gateway-s3x-object.go:338]: Ok for some reason I thought we were taking the prefix but I should have paid more attentioned to `prefixLen:` :+1:

**Pull request merged**